src/main/java/org/opentripplanner/routing/algorithm/strategies/PriceBasedRemainingWeightHeuristic.java

```diff
@@ -0,0 +1,95 @@
+package org.opentripplanner.routing.algorithm.strategies;
+
+import org.opentripplanner.common.geometry.SphericalDistanceLibrary;
+import org.opentripplanner.routing.algorithm.costs.CostFunction;
+import org.opentripplanner.routing.core.RoutingRequest;
+import org.opentripplanner.routing.core.State;
+import org.opentripplanner.routing.core.vehicle_sharing.VehicleDescription;
+import org.opentripplanner.routing.core.vehicle_sharing.VehiclePricingPackage;
+import org.opentripplanner.routing.edgetype.StreetEdge;
+import org.opentripplanner.routing.graph.Edge;
+import org.opentripplanner.routing.graph.Vertex;
+
+import java.math.BigDecimal;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+
+public class PriceBasedRemainingWeightHeuristic implements RemainingWeightHeuristic {
+
+    private static final double DEFAULT_PRICE_COST_WEIGHT = 1;
+    private static final double DEFAULT_ORIGINAL_COST_WEIGHT = 0.01;
+
+    private double lat;
+    private double lon;
+    private double priceCostWeight;
+    private double originalCostWeight;
+    private EuclideanRemainingWeightHeuristic originalHeuristic;
+
+    public PriceBasedRemainingWeightHeuristic(Map<CostFunction.CostCategory, Double> costCategoryWeights){
+        priceCostWeight = Optional.ofNullable(costCategoryWeights.get(CostFunction.CostCategory.PRICE_ASSOCIATED)).orElse(DEFAULT_PRICE_COST_WEIGHT);
+        originalCostWeight = Optional.ofNullable(costCategoryWeights.get(CostFunction.CostCategory.ORIGINAL)).orElse(DEFAULT_ORIGINAL_COST_WEIGHT);
+        originalHeuristic = new EuclideanRemainingWeightHeuristic();
+    }
+
+    @Override
+    public void initialize(RoutingRequest request, long abortTime) {
+        Vertex target = request.rctx.target;
+        lat = target.getLat();
+        lon = target.getLon();
+        originalHeuristic.initialize(request, abortTime);
+    }
+
+    @Override
+    public double estimateRemainingWeight(State s) {
+        Vertex currentLocation = s.getVertex();
+        double remainingDistance = SphericalDistanceLibrary.distance(currentLocation.getLat(), currentLocation.getLon(), lat, lon);
+
+        double speed;
+        int remainingTime;
+        double estimatedFuturePrice = 0;
+        Edge backEdge = s.getBackEdge();
+        if (Objects.nonNull(s.getCurrentVehicle()) && Objects.nonNull(backEdge) && backEdge instanceof StreetEdge ) {
+            speed = s.getCurrentVehicle().getMaxSpeedInMetersPerSecond((StreetEdge) backEdge);
+            remainingTime = (int) (remainingDistance / speed);
+            estimatedFuturePrice = chooseBestFuturePrice(s.getCurrentVehicle(),
+                    s.getTimeTraversedInCurrentVehicleInSeconds(), s.getDistanceTraversedInCurrentVehicle(),
+                    remainingTime, remainingDistance);
+        }
+
+        return priceCostWeight * estimatedFuturePrice + originalCostWeight * originalHeuristic.estimateRemainingWeight(s);
```

Comment:
Tak się zastanawiam, czy w takim wypadku możemy w ogóle szacować koszt dotarcia do celu? Bo przecież najtaniej byłoby oddać nasz pojazd już teraz i iść do celu pieszo. Tutaj poprawnym oszacowaniem byłaby pewnie minimalizacja sumy ceny i wagi (bo pojazdy są dużo szybsze od chodzenia pieszo, ale są też drogie). Robienie takich wyliczeń byłoby jakieś kuriozalnie trudne, bo to nie są dwie funkcje liniowe - z jednej strony mamy bardzo skomplikowane wyliczenia ceny, a z drugiej mamy maksymalny dystans pieszy. @MironSz możliwe, że coś namieszałem, więc może Ty na to zerknij

Reply:
Zgadzam się, że ta linia jest bardzo niefortunna. Oddanie samochodu zeruje nam pierwszą część tej sumy a zostawia nieruszoną drugą. W efekcie stan po wysiąściu z auta jest preferowany nad analogiczny stan  w aucie.  Moje trzy propozycje to:
1. Olanie części estimatedFuturePrice
2. estimateRemainingWeight zależy od prędkości (obecnie domyślnie samochodu). Można uzależnić tę estymację od obecnego pojazdu
3. Jak wyżej ale zrobić minimum po wszystkich trybach.

Generalnie każdą z tych możliwości trzeba mocno wytestować pod względem przyśpieszenia i poprawności. Jak ja robię zmiany w heurystyce to dodaję opcje heurystyki do requesta i korzystam z skryptu w drugim repo.